kafka-client/src/main/java/dev/responsive/kafka/internal/db/WriteBatch.java

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ *  * Copyright 2023 Responsive Computing, Inc.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package dev.responsive.kafka.internal.db;
+
+import dev.responsive.kafka.internal.db.partitioning.TablePartitioner;
+import dev.responsive.kafka.internal.stores.RemoteWriteResult;
+import dev.responsive.kafka.internal.utils.Result;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+/**
+ * A class for holding and coordinating the pending writes and other metadata involved
+ * in a flush to the remote table
+ */
+public class WriteBatch<K extends Comparable<K>, P> {
```

Comment:
I think this would be cleaner as just a POJO that encapsulates a batch that's ready to be written, with all the code that actually executes the write moved to `WriteBatcher`. I personally would find this easier to make sense of (because the way it's currently chopped up I'm not sure what's "supposed" to be in `WriteBatcher` vs `WriteBatch`, but there's some more objective benefits too:
    - you could also get rid of the `closed` check below.
    - you wind up with code that's less easy to use incorrectly. Right now, something outside this package could create a WriteBatch, which also makes all these `writer.delete` / `writer.insert` calls without going through the proper flush process. If `WriteBatch` was a POJO you could create it sure, but you'd have to pass it to `WriteBatcher` (which does the write using the right protocol) to actually do anything.

Reply:
Cleaned it up so that this class only keeps track of the actual writers for the batch, rather than handling the writes themselves. The main goal of this class was to encapsulate the writers for a single given batch, but since we have to iterate over all the buffered writes to build up the set of writers, I figured we may as well stage those writes as well. 

We now have to iterate over the batch twice now (first to build up the set of writers based on the keys, then later to issue the actual `writer.insert/delete` operations) but that's not too big a price if the outcome feels cleaner (does it?)